### ISSUE TITLE
ParagraphTextObject-VisualMode Problems

### ISSUE THREAD

**Issue by solde9** [vrapper/vrapper#725]
These problems are more like the vim-compatible kind-of. 

###### 1) Extending selection
"a paragraph"/"inner paragraph" text objects won't extend current selection. Instead, it selects a new paragraph [starting from the cursor](https://github.com/vrapper/vrapper/blob/5278ff6065930eaf6c707840080af0c6a89f66e7/net.sourceforge.vrapper.core/src/net/sourceforge/vrapper/vim/commands/motions/ParagraphMotion.java#L112).

Example:
Initial state:

![image](https://cloud.githubusercontent.com/assets/2660849/13205302/29ebf7dc-d8c4-11e5-8427-ac0baac230b8.png)

After hitting `ap`:

![image](https://cloud.githubusercontent.com/assets/2660849/13205309/5ab10434-d8c4-11e5-806c-5968cb7fed8e.png)

Expected selection:

![image](https://cloud.githubusercontent.com/assets/2660849/13205320/8f25407c-d8c4-11e5-910e-06003b3a6e93.png)

###### 2) Direction of Selection

It extends only forward, not matter what direction the current selection has.

Example:
Initial state:

![image](https://cloud.githubusercontent.com/assets/2660849/13205363/537fa642-d8c5-11e5-9e8a-3d4b7db9d757.png)

After hitting `ap`:

![image](https://cloud.githubusercontent.com/assets/2660849/13205366/62908a02-d8c5-11e5-8b34-3c47d2dba0f2.png)

Expected selection:

![image](https://cloud.githubusercontent.com/assets/2660849/13205378/9d0768b8-d8c5-11e5-9c3f-8853aa4c6842.png)

**Comment by albertdev** [comment]
I think we're hitting the same issues as #474 where the textobject implementations lack the necessary context to extend the selection because they don't receive a `Selection` as input.

**Comment by keforbes** [comment]
I decided to look into this issue since it also affects SentenceMotion, which I just implemented in #715.  I think the defect originates in `SelectTextObjectCommand.execute()`.  That method executes the text object and updates the selection with the resulting range.  Part of the calculation looks at the current selection but that selection isn't used when determining the new selection.  I thought I could simply add a `start = Math.min(old, new)`, `end = Math.max(old, new)` to expand the current selection and include the new range.  This *almost* works.  I'm hitting an issue with the inclusive/exclusive behavior though.

We have multiple implementations of our `Selection` class and I need to modify whatever `Selection` instance is used so I can include this new range.  The only method available to modify an existing selection instance is `Selection.reset()`, which recalculates the start/end.  Unfortunately, I think it also recalculates the inclusive/exclusive behavior, which the `region` already calculated for us.  This gives me an off-by-one issue.  If I take the new `end` as calculated, I include the first character in the next sentence/paragraph rather than selecting *up to* the boundary.  If I try anticipating that behavior by stepping back one character, the cursor refuses to jump to the next line.  I think the issue here is that each selection type (rectangle, linewise, text) has its own inclusive/exclusive behavior.

Now, I *could* probably check the `instanceof` the current selection class and behave accordingly, but that seems like such a hack.  I probably just need to spend more time with this but I wanted to share my progress in case I'm missing something obvious.

For posterity, here's what I'm currently playing with in that method:
```
Position newStart = oldSelection.getStart().compareTo(region.getStart()) < 0 ? oldSelection.getStart() : region.getStart();
Position newEnd = oldSelection.getEnd().compareTo(region.getEnd()) > 0 ? oldSelection.getEnd() : region.getEnd();
selection = selection.reset(editorAdaptor, newStart, newEnd);
```

**Comment by albertdev** [comment]
>  I think the issue here is that each selection type (rectangle, linewise, text) has its own inclusive/exclusive behavior.

Correct; visual mode knows about inclusive/exclusive whereas visual line mode is *always* inclusive. Why are you relying on that behavior though and why are you modifiying the range in `SelectTextObjectCommand` and not in your custom text object?

Also note that the simple `TEXT` content type won't trigger a change of text ranges but the constructor for `LineWiseSelection` would. As a workaround it seems a text object with content type `LINES` can immediately return a `LineWiseSelection` to return the proper result.

**Comment by keforbes** [comment]
I was looking at modifying things in `SelectTextObjectCommand` because this defect is applicable to all text objects, not just paragraph and sentence.  Paragraph and sentence have the most sensible use case for this but we're still blowing away any previous selections when a text object is used in visual mode.

**Comment by albertdev** [comment]
I agree that the current behaviour might not be correct either way, it's just that I'm questioning whether `SelectTextObjectCommand` is the best place to do it and if we don't need to replace the `TextObject` interface to solve the core problem (sadly breaking all text objects in the process).

`TextObject` always returns a `TextRange` with offsets as bounds for some operation. This works great for `3caw` or `vap`, but I think it can no longer work if you see the following Vim behavior:

Start with this text and on the first line: 

> Vim has lots of little one-offs. Sentences for example. Selecting in line mode and then
> jumping to the next "all" sentence will jump out of line mode.

* Start on the first line and press `0fc`.
* Enter visual line mode.
* Press `as`. The selection will *shrink* to wrap the second sentence.
* Press `as` again. The selection will extend to the third sentence on the second line, staying in visual mode.

Slightly different scenario:

* Start on the first line and press `0fc`.
* Press `as`, visual mode will be entered and the second sentence will be selected.
* Press `V` to enter visual line mode.
* Press `as`, the caret will be moved to include the third sentence but Vim will stay in visual line mode.

Now what I *think* is happening here is that Vim looks at the current selection's `from` and `to` offsets. The selected text object pushes the `from` and `to` offsets outwards and then decides whether to stick to the current visual mode or to change to a different mode.

Our implementation of text objects might thus need to be changed similar to the way selections are now based on `from` and `to` fields rather than the raw offsets of their boundaries, while still keeping some way of passing raw boundaries for the Operations to do their work. I haven't figured out what would be the best interface with the least amount of refactoring.

**Comment by solde9** [comment]
An interesting behavior is that it pushes `from` and `to` outwards only when they are the same. If they are different, It only pushes the `RightBound`, as in this scenario (reusing the same example):

- Start on the first line and press `0fc`.
- Press `vl`, visual mode will be entered and the cursor will move one character to the right. 
- Press `as`. The selection now spans the text `ces for example. `
